drivers: pwm: pwm_mcux_sctimer: Fix stopping PWM

morandg opened the pull request:
When the pulse was set to 0, the PWM controller was never correctly stopped and used the last configured pulse instead.

We now configure the device and apply the new duty cycle everytime.

Signed-off-by: Guy Morand <guy.morand@bytesatwork.ch>

decsny commented:
#51392 adds sctimer

mmahadevan108 commented on drivers/pwm/pwm_mcux_sctimer.c:
```diff
@@ -60,7 +60,7 @@ static int mcux_sctimer_pwm_set_cycles(const struct device *dev,
 
 	duty_cycle = 100 * pulse_cycles / period_cycles;
-
+	data->channel[channel].dutyCyclePercent = duty_cycle;
```
This change is not needed as `channel[channel].dutyCyclePercent` is only used by `SCTIMER_SetupPwm()` function

mmahadevan108 commented on drivers/pwm/pwm_mcux_sctimer.c:
```diff
@@ -102,7 +100,6 @@ static int mcux_sctimer_pwm_set_cycles(const struct device *dev,
 		SCTIMER_StopTimer(config->base, kSCTIMER_Counter_U);
 
 		LOG_DBG("SETUP dutycycle to %u\n", duty_cycle);
-		data->channel[channel].dutyCyclePercent = duty_cycle;
```
This can be left as it due to the same reason as above.

mmahadevan108 commented:
@morandg thank you for your PR. Can you please share the steps you used to see the failure?

morandg commented:
Thanks @mmahadevan108 for your feedback. I was using a vibration motor. Here are the steps (pseudo code) to reproduce the problem I had:
```
pwm_set_pulse_dt(pwm, pwm->period);      /* Vibrate 100% */
pwm_set_pulse_dt(pwm, 0);                /* Still Vibrate 100% */
pwm_set_pulse_dt(pwm, pwm->period / 2);  /* Vibrate 50% */
pwm_set_pulse_dt(pwm, 0);                /* Still Vibrate 50% */
```
With this patch, I was able to stop the motor with ```pwm_set_pulse_dt(pwm, 0); ```

morandg commented on drivers/pwm/pwm_mcux_sctimer.c:
```diff
@@ -77,8 +77,6 @@ static int mcux_sctimer_pwm_set_cycles(const struct device *dev,
 		if (data->period_cycles[channel] != 0) {
 			SCTIMER_StartTimer(config->base, kSCTIMER_Counter_U);
 		}
```
Excellent, thanks for your feedback! Unfortunately, I'm not able to test right now, it will be done by the beginning of the next week.

morandg committed:
drivers: pwm: pwm_mcux_sctimer: Fix stopping PWM

By Starting the timer when period was set to 0, the SCTimer was giving
control of the pin back to the timer and over-riding the value written
to base->OUTPUT register. Consequently, the PWM timer was never
stopped and still using the previously configured period.

The PWM now correctly stops when setting the period to 0.

Signed-off-by: Guy Morand <guy.morand@bytesatwork.ch>

morandg commented:
OK @mmahadevan108, I tested you change and can confirm it fixes my issue. I changed the commit message consequently, I'm just ashamed, you deserve the honor :wink: ! Thanks for taking time in looking into this!

mmahadevan108 commented:
> OK @mmahadevan108, I tested you change and can confirm it fixes my issue. I changed the commit message consequently, I'm just ashamed, you deserve the honor 😉 ! Thanks for taking time in looking into this!

@morandg, thank you for finding this issue and for the fix.
